File: src/windows/AgentHaloBinaryStaging.cs
using System;
using System.IO;

namespace CodexHalo
{
    /// <summary>
    /// Atomic install helpers for staged hook binaries under %USERPROFILE%\.agent-halo.
    /// Mirrors macOS upgrade invariants: never leave a advertised path missing.
    /// </summary>
    internal static class AgentHaloBinaryStaging
    {
        public static void StageExecutable(string source, string destination)
        {
            if (String.IsNullOrEmpty(source) || !File.Exists(source))
            {
                throw new FileNotFoundException("source executable missing", source);
            }

            string parent = Path.GetDirectoryName(destination);
            if (!String.IsNullOrEmpty(parent))
            {
                Directory.CreateDirectory(parent);
            }

            string temp = Path.Combine(
                parent ?? Path.GetTempPath(),
                ".staging-" + Path.GetFileName(destination) + "-" +
                Guid.NewGuid().ToString("N"));
            try
            {
                File.Copy(source, temp, true);
                if (File.Exists(destination))
                {
                    // Atomic replace on same volume when possible.
                    try
                    {
                        File.Replace(temp, destination, null);
                        temp = null;
                        return;
                    }
                    catch
                    {
                        // Fall through: overwrite copy (still no delete-first gap
                        // for the destination path itself).
                        File.Copy(temp, destination, true);
                        return;
                    }
                }
                File.Move(temp, destination);
                temp = null;
            }
            finally
            {
                if (temp != null)
                {
                    try { File.Delete(temp); }
                    catch { }
                }
            }
        }

        /// <summary>
        /// Stage current AgentHalo.exe to the preferred bin\status-hook.exe only.
        /// Configurators rewrite settings to this path on launch.
        /// </summary>
        public static void StageStatusHookEverywhere(
            string sourceExecutable,
            string userProfile = null)
        {
            StageExecutable(sourceExecutable, AgentHaloPaths.StatusHookExe(userProfile));
        }

        /// <summary>
        /// After settings point at bin\status-hook.exe, remove unreferenced
        /// root-level AgentHaloHook.exe leftovers.
        /// </summary>
        public static void ScrubUnreferencedLegacyBinaries(string userProfile = null)
        {
            try
            {
                string legacy = AgentHaloPaths.LegacyAgentHaloHookExe(userProfile);
                if (!File.Exists(legacy))
                {
                    return;
                }
                string home = String.IsNullOrEmpty(userProfile)
                    ? Environment.GetFolderPath(Environment.SpecialFolder.UserProfile)
                    : userProfile;
                string settings = Path.Combine(home, ".claude", "settings.json");
                if (File.Exists(settings))
                {
                    string text = File.ReadAllText(settings);
                    if (text.IndexOf("AgentHaloHook.exe", StringComparison.OrdinalIgnoreCase) >= 0)
                    {
                        return;
                    }
                }
                File.Delete(legacy);
                SettingsStorage.Log(
                    "AgentHaloBinaryStaging: removed unreferenced " + legacy);
            }
            catch (Exception ex)
            {
                SettingsStorage.Log(
                    "AgentHaloBinaryStaging: legacy scrub failed: " + ex.Message);
            }
        }

        /// <summary>
        /// Extract the executable token from a hook command, handling quotes.
        /// </summary>
        public static string CommandExecutablePath(string command)
        {
            if (String.IsNullOrWhiteSpace(command))
            {
                return String.Empty;
            }
            string trimmed = command.Trim();
            if (trimmed.StartsWith("\"", StringComparison.Ordinal))
            {
                int end = trimmed.IndexOf('"', 1);
                if (end > 1)
                {
                    return trimmed.Substring(1, end - 1);
                }
            }
            int space = trimmed.IndexOf(' ');
            return space < 0 ? trimmed : trimmed.Substring(0, space);
        }

        public static bool CommandPointsToLiveExecutable(string command)
        {
            string path = CommandExecutablePath(command);
            if (String.IsNullOrEmpty(path))
            {
                return false;
            }
            try
            {
                return File.Exists(path);
            }
            catch
            {
                return false;
            }
        }
    }
}

File: src/macos/Sources/AgentHaloCore/AgentHaloBinaryStaging.swift
import Foundation

/// Atomic install helpers for staged hook / statusline binaries under
/// ``~/.agent-halo/bin``.
///
/// Upgrade sequence (prevent exit 127):
/// 1. Atomically stage the **preferred** binary under ``bin/`` (never delete-first).
/// 2. Configurators rewrite Claude/Grok settings so every Agent Halo command
///    points at that preferred path.
/// 3. Only then may leftover root-level legacy names be removed if nothing
///    references them.
public enum AgentHaloBinaryStaging {
    /// Copy `source` onto `destination` via a same-directory temp file, then
    /// replace. The destination path stays executable for concurrent hook runs
    /// when it already existed.
    public static func stageExecutable(
        from source: URL,
        to destination: URL,
        fileManager: FileManager = .default
    ) throws {
        try fileManager.createDirectory(
            at: destination.deletingLastPathComponent(),
            withIntermediateDirectories: true,
            attributes: [.posixPermissions: 0o700]
        )

        let temp = destination
            .deletingLastPathComponent()
            .appendingPathComponent(".staging-\(destination.lastPathComponent)-\(UUID().uuidString)")
        defer { try? fileManager.removeItem(at: temp) }

        if fileManager.fileExists(atPath: temp.path) {
            try fileManager.removeItem(at: temp)
        }
        try fileManager.copyItem(at: source, to: temp)
        try fileManager.setAttributes(
            [.posixPermissions: 0o755],
            ofItemAtPath: temp.path
        )

        if fileManager.fileExists(atPath: destination.path) {
            _ = try fileManager.replaceItemAt(destination, withItemAt: temp)
        } else {
            try fileManager.moveItem(at: temp, to: destination)
        }
        try fileManager.setAttributes(
            [.posixPermissions: 0o755],
            ofItemAtPath: destination.path
        )
    }

    /// Stage only the preferred ``bin/status-hook`` path.
    public static func stageStatusHook(
        from bundled: URL,
        homeDirectory: URL,
        fileManager: FileManager = .default
    ) throws {
        let paths = AgentHaloPaths(homeDirectory: homeDirectory)
        try stageExecutable(from: bundled, to: paths.statusHook, fileManager: fileManager)
    }

    /// Stage only the preferred ``bin/statusline-proxy`` path.
    public static func stageStatuslineProxy(
        from bundled: URL,
        homeDirectory: URL,
        fileManager: FileManager = .default
    ) throws {
        let paths = AgentHaloPaths(homeDirectory: homeDirectory)
        try stageExecutable(from: bundled, to: paths.statuslineProxy, fileManager: fileManager)
    }

    /// Backward-compatible name used by bootstrap / older call sites.
    public static func stageStatusHookEverywhere(
        from bundled: URL,
        homeDirectory: URL,
        fileManager: FileManager = .default
    ) throws {
        try stageStatusHook(from: bundled, homeDirectory: homeDirectory, fileManager: fileManager)
    }

    public static func stageStatuslineProxyEverywhere(
        from bundled: URL,
        homeDirectory: URL,
        fileManager: FileManager = .default
    ) throws {
        try stageStatuslineProxy(from: bundled, homeDirectory: homeDirectory, fileManager: fileManager)
    }

    /// First path token of a hook command (`"/path/bin args"` → `"/path/bin"`).
    public static func commandExecutablePath(_ command: String) -> String {
        let trimmed = command.trimmingCharacters(in: .whitespacesAndNewlines)
        if trimmed.hasPrefix("\""),
           let end = trimmed.dropFirst().firstIndex(of: "\"") {
            return String(trimmed[trimmed.index(after: trimmed.startIndex)..<end])
        }
        guard let space = trimmed.firstIndex(of: " ") else {
            return trimmed
        }
        return String(trimmed[..<space])
    }

    public static func commandPointsToLiveExecutable(
        _ command: String?,
        fileManager: FileManager = .default
    ) -> Bool {
        guard let command, !command.isEmpty else { return false }
        let path = commandExecutablePath(command)
        guard !path.isEmpty else { return false }
        return fileManager.isExecutableFile(atPath: path)
    }

    public static func commandReferencesPath(_ command: String?, _ path: String) -> Bool {
        guard let command else { return false }
        let exe = commandExecutablePath(command)
        return exe == path || command == path || command.hasPrefix(path + " ") || command.contains(path)
    }

    /// After settings have been rewritten to preferred paths, drop root-level
    /// legacy binaries that nothing still references.
    public static func scrubUnreferencedLegacyBinaries(
        homeDirectory: URL,
        fileManager: FileManager = .default
    ) {
        let paths = AgentHaloPaths(homeDirectory: homeDirectory)
        let referenced = referencedExecutablePaths(homeDirectory: homeDirectory, fileManager: fileManager)

        for legacy in [paths.legacyStatusHook, paths.legacyStatuslineProxy] {
            guard fileManager.fileExists(atPath: legacy.path) else { continue }
            if referenced.contains(legacy.path) {
                continue
            }
            do {
                try fileManager.removeItem(at: legacy)
                AgentHaloLogger.log("AgentHaloBinaryStaging: removed unreferenced \(legacy.path)")
            } catch {
                AgentHaloLogger.log(
                    "AgentHaloBinaryStaging: failed to remove \(legacy.path): \(error)"
                )
            }
        }

        // Emergency Grok-local binary from the 127 incident.
        let grokOrphan = homeDirectory
            .appendingPathComponent(".grok/hooks/agent-halo-status-hook")
        if fileManager.fileExists(atPath: grokOrphan.path),
           !referenced.contains(grokOrphan.path) {
            try? fileManager.removeItem(at: grokOrphan)
        }
    }

    private static func referencedExecutablePaths(
        homeDirectory: URL,
        fileManager: FileManager
    ) -> Set<String> {
        var refs = Set<String>()
        let candidates = [
            homeDirectory.appendingPathComponent(".claude/settings.json"),
            homeDirectory.appendingPathComponent(".grok/hooks/agent-halo-status.json"),
        ]
        for url in candidates {
            guard let data = try? Data(contentsOf: url),
                  let text = String(data: data, encoding: .utf8) else {
                continue
            }
            // Cheap scan: collect absolute path-like tokens that look like our binaries.
            for name in [
                "claude-code-status-hook",
                "claude-code-statusline-proxy",
                "status-hook",
                "statusline-proxy",
                "agent-halo-status-hook",
            ] {
                if text.contains(name) {
                    // If the file mentions the legacy basename, treat legacy path as referenced.
                    let paths = AgentHaloPaths(homeDirectory: homeDirectory)
                    switch name {
                    case "claude-code-status-hook":
                        refs.insert(paths.legacyStatusHook.path)
                    case "claude-code-statusline-proxy":
                        refs.insert(paths.legacyStatuslineProxy.path)
                    case "status-hook":
                        refs.insert(paths.statusHook.path)
                    case "statusline-proxy":
                        refs.insert(paths.statuslineProxy.path)
                    case "agent-halo-status-hook":
                        refs.insert(
                            homeDirectory
                                .appendingPathComponent(".grok/hooks/agent-halo-status-hook").path
                        )
                    default:
                        break
                    }
                }
            }
        }
        return refs
    }
}
